### PR TITLE
Add validate logic for precision

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

Ensures that we validate the precision type when setting the precision plugin. Followup from #9049

```python
Trainer(precision="x")
``` 

```
pytorch_lightning.utilities.exceptions.MisconfigurationException: Precision "x" is invalid. Allowed precision values: (16, 32, 64, 'bf16')
```

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 1]
```diff
@@ -546,7 +546,18 @@ def is_using_torchelastic(self) -> bool:
         )
         return TorchElasticEnvironment.is_using_torchelastic()
 
+    def _validate_precision_type(self) -> None:
+        """
+        Ensures that the set precision type by the user is valid.
+        """
+        valid_types = (16, 32, 64, "bf16")
```
Comment (kaushikb11) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:553]: Maybe create a `PrecisionType` Enum?

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 2]
```diff
@@ -546,7 +546,18 @@ def is_using_torchelastic(self) -> bool:
...
+        if self.precision not in valid_types:
+            raise MisconfigurationException(
+                f"Precision {self.precision} is invalid. Allowed precision values: {valid_types}"
+            )
+
     def select_precision_plugin(self) -> PrecisionPlugin:
+        self._validate_precision_type()
```
Comment (kaushikb11) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:560]: Would be better to add it to the Connector's constructor?

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 3]
```diff
@@ -546,7 +546,18 @@ def is_using_torchelastic(self) -> bool:
         )
         return TorchElasticEnvironment.is_using_torchelastic()
 
+    def _validate_precision_type(self) -> None:
+        """
+        Ensures that the set precision type by the user is valid.
```
Comment (kaushikb11) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:551]:
```suggestion
        Ensures that the set precision type passed by the user is valid.
```

File: pytorch_lightning/trainer/connectors/accelerator_connector.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:553]: I think we should support bfloat16 too as user might try this option first coming from PyTorch documentation after reading updates.

Comment (kaushikb11) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:553]: Agreed!

Comment (justusschock) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:553]: And I think we should also add `precision='amp'` and deprecate `precision=16` for now and as soon as we remove it, we can add true half precision training instead of amp.

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 4]
```diff
@@ -148,6 +149,8 @@ def __init__(
         self.plugins = plugins
 
         self._validate_accelerator_and_devices()
+        self._validate_precision_type()
```
Comment (kaushikb11) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:152]:
```suggestion
        self._validate_precision_type(precision)
```

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 5]
```diff
@@ -546,6 +549,16 @@ def is_using_torchelastic(self) -> bool:
...
         return TorchElasticEnvironment.is_using_torchelastic()
 
+    def _validate_precision_type(self) -> None:
+        """
+        Ensures that the set precision type passed by the user is valid.
+        """
+
+        if not PrecisionType.supported_type(self.precision):
+            raise MisconfigurationException(
+                f"Precision {self.precision} is invalid. Allowed precision values: {PrecisionType.supported_types()}"
```
Comment (kaushikb11) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:559]:
```suggestion
	@staticmethod
    def _validate_precision_type(precision: Union[int, str]) -> None:
        """
        Ensures that the set precision type passed by the user is valid.
        """

        if not PrecisionType.supported_type(precision):
            raise MisconfigurationException(
                f"Precision {precision} is invalid. Allowed precision values: {PrecisionType.supported_types()}"
```

**Review by kaushikb11 — APPROVED** [review]
LGTM!

File: pytorch_lightning/utilities/enums.py [hunk 6]
```diff
@@ -49,6 +49,29 @@ class AMPType(LightningEnum):
...
+    FLOAT = 32
+    FULL = 64
+    BFLOAT = "bf16"
+
+    @staticmethod
+    def supported_type(precision: Union[str, int]) -> bool:
+        return any(x == precision for x in PrecisionType)
+
+    @staticmethod
+    def supported_types() -> List[str]:
```
Comment (kaushikb11) [inline pytorch_lightning/utilities/enums.py:71]:
```suggestion
    def supported_types() -> List[Union[int, str]]:
```

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 7]
```diff
@@ -601,7 +614,7 @@ def select_precision_plugin(self) -> PrecisionPlugin:
                 log.info("Using APEX 16bit precision.")
                 return ApexMixedPrecisionPlugin(self.amp_level)
 
-        raise NotImplementedError("We only support precisions 64, 32 and 16!")
+        raise NotImplementedError(f"We only support precision: {PrecisionType.supported_types()}")
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:617]: this won't ever be raised since you already check this in `_validate_precision_type`, correct?

Inline the code of `_validate_precision_type` here?

File: pytorch_lightning/utilities/enums.py [hunk 8]
```diff
@@ -49,6 +49,29 @@ class AMPType(LightningEnum):
...
+
+    >>> PrecisionType.HALF == 16
+    True
+    >>> PrecisionType.HALF in (16, "16")
+    True
+    """
+
+    HALF = 16
+    FLOAT = 32
+    FULL = 64
```
Comment (carmocca) [inline pytorch_lightning/utilities/enums.py:63]: Are you sure this works well?

`LightningEnum` subclasses `str, Enum` so the int values here might break things.

Convert them to str?

File: pytorch_lightning/utilities/enums.py (hunk 6 above)
Comment (SeanNaren) [reply pytorch_lightning/utilities/enums.py:71]: This is never true, because I think the enum will always return a string

File: pytorch_lightning/utilities/enums.py (hunk 8 above)
Comment (SeanNaren) [reply pytorch_lightning/utilities/enums.py:63]: Seemed to work fine (the doc test explicitly tests that an int comparison works), have converted them to string though for explicitness!

File: pytorch_lightning/trainer/connectors/accelerator_connector.py (hunk 7 above)
Comment (SeanNaren) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:617]: Yep exactly, I'm unsure what to do here... any suggestions?

**Review by tchaton — APPROVED** [review]
Looks great !

File: pytorch_lightning/utilities/enums.py [hunk 9]
```diff
@@ -49,6 +49,29 @@ class AMPType(LightningEnum):
...
+    FULL = "64"
+    BFLOAT = "bf16"
+
+    @staticmethod
+    def supported_type(precision: Union[str, int]) -> bool:
+        return any(x == precision for x in PrecisionType)
+
+    @staticmethod
+    def supported_types() -> List[str]:
+        return [x.value for x in PrecisionType]
```
Comment (carmocca) [inline pytorch_lightning/utilities/enums.py:72]: This method is not necesary, `list(PrecisionType)` could be used instead

**Review by carmocca — APPROVED** [review]

**Review by awaelchli — COMMENTED** [review]
LGTM
